### PR TITLE
Expression language in Python

### PR DESCRIPTION
**Pull request by jigold**
Here's what I have now. Higher-level feedback appreciated. Note there are 4 hidden methods to convert between old interface and new.

`VariantDataset._to_new_variant_dataset`
`NewVariantDataset._to_old_variant_dataset`
`KeyTable._to_new_keytable`
`NewKeyTable._to_old_keytable`

### CODE REVIEW

File: python/hail/__init__.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 from hail.representation import *
 from hail.context import HailContext
 from hail.dataset import VariantDataset
-from hail.expr import *
+from hail.htypes import *
```
Comment (cseed) [inline python/hail/__init__.py]: I don't like putting h[ail] as a prefix on things already in the hail package.  I prefer `typ` or `type_` or `types` or something.  The Python style guide:

https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

says, "single_trailing_underscore_: used by convention to avoid conflicts with Python keyword, e.g. `Tkinter.Toplevel(master, class_='ClassName')`".

File: python/hail/dataset.py [hunk 2]
```diff
@@ -5809,4 +5809,9 @@ def make_table(self, variant_expr, genotype_expr, key=[], separator='.'):
                                 jarray(Env.jvm().java.lang.String, wrap_to_list(key)), separator)
         return KeyTable(self.hc, jkt)
 
+    def _to_new_variant_dataset(self):
```
Comment (cseed) [inline python/hail/dataset.py]: OK, I think you should make a second module for the new python expr interface.  It should be self-contained and there should conversion routines between the old and new setups.  Call the new one `newhail` or `hail2` or `hailpyexpr` or something.  (I'll call it `hail2`.)  There should be a `hail2.HailContext` that returns `hail2.KeyTable`, etc.

`hail.VariantDataset` should have `to_hail2` and `hail2.VariantDataset` should similarly have `to_hail1` (or `to_oldhail` or whatever.)

File: python/hail/expr/__init__.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import hail.expr
+from hail.expr.functions import *
+from hail.expr.column import Column, NumericColumn
+from hail.expr.keytable import GroupedKeyTable, NewKeyTable
+from hail.expr.dataset import NewVariantDataset
+
+__all__ = ['NewKeyTable',
```
Comment (cseed) [inline python/hail/expr/__init__.py]: Can you add the new classes to the docs?  Functions with no docstrings will just show up with function names/args, right?  Even that is probably useful for guinea pigs for the new interface.

Comment (cseed) [reply python/hail/expr/__init__.py]: Question about this `__all__`.  Generally, I'd hope people could do `from hail2 import *` and they'd have everything they'd need to use the Hail interface, but that doesn't seem to be the case here because a ton of functionality comes through functions, but those aren't exported.  Is the intention to have functions in the default import or to do something like `import hail2.functions as hf` and then `hf.chisq(a, b, c, d)`.

There is good precedent for choosing a short (2-letter) import (e.g., np, tf, pd, ...) and accessing everything through that.  Something like:

```
import hail2 as hl
hc = hl.HailContext()
... hl.chisq(a, b, c, d) ...
```
What do you think?

File: python/hail/expr/column.py [hunk 4]
```diff
@@ -0,0 +1,1192 @@
...
+    elif isinstance(arg, Locus):
+        return "Locus(\"" + str(arg) + "\")"
+    elif isinstance(arg, Interval):
+        return "Interval(\"" + str(arg) + "\")"
+    elif isinstance(arg, Struct):
+        return "{" + ", ".join([k + ":" + to_expr(v) for k, v in arg._attrs.iteritems()]) + "}"  # FIXME: Struct constructor should take kwargs (ordered in Python 3.6)
+    elif callable(arg):
+        return arg
+    else:
+        return str(arg)
```
Comment (cseed) [inline python/hail/expr/column.py]: Do you know what might go into this case?  For a case this long and complicated, I'd probably prefer an error here so we know explicit what we're missing.

File: python/hail/expr/column.py [hunk 5]
```diff
@@ -0,0 +1,1192 @@
...
+    elif isinstance(arg, str):
+        return "\"" + arg + "\""
+    elif isinstance(arg, bool):
+        return "true" if arg else "false"
+    elif isinstance(arg, long):
+        return "{}.toInt64()".format(arg)
+    elif isinstance(arg, list):
+        return "[" + ", ".join([to_expr(a) for a in arg]) + "]"
+    elif isinstance(arg, set):
+        return "let xfsjd = [" + ", ".join([to_expr(a) for a in arg]) + "] in xfsjd.toSet()"
```
Comment (cseed) [inline python/hail/expr/column.py]: Hmm, a bit dicy.  We should add a function on HailContext (which you can get through the Env) to keep a counter and generate unique symbols of the form `__gen_N` say.  This isn't foolproof but probably good enough for this prototype, and we can fix it properly when we're generating the AST directly.

File: python/hail/expr/column.py (hunk 4 above)
Comment (cseed) [reply python/hail/expr/column.py]: Same with other big case analyses.

File: python/hail/expr/dataset.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
+from __future__ import print_function  # Python 2 and 3 print compatibility
+
+import warnings
+
+from decorator import decorator
+from hail.expr.column import *
+
+
+warnings.filterwarnings(module=__name__, action='once')
```
Comment (cseed) [inline python/hail/expr/dataset.py]: Are we getting a ton of warnings?  What for?

File: src/main/scala/is/hail/expr/Parser.scala [hunk 7]
```diff
@@ -400,7 +400,7 @@ object Parser extends JavaTokenParsers {
     }
 
   def unary_expr: Parser[AST] =
-    rep(withPos("-" | "!" | "**")) ~ exponent_expr ^^ { case lst ~ rhs =>
+    rep(withPos("-" | "+" | "!")) ~ exponent_expr ^^ { case lst ~ rhs =>
```
Comment (cseed) [inline src/main/scala/is/hail/expr/Parser.scala]: This should go in a separate PR!

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 8]
```diff
@@ -1563,7 +1563,7 @@ object FunctionRegistry {
   registerMethod("union", (x: Set[Any], a: Set[Any]) => x ++ a, "Returns the union of this Set and Set `a`.")(setHr(TTHr), setHr(TTHr), setHr(TTHr))
   registerMethod("intersection", (x: Set[Any], a: Set[Any]) => x & a, "Returns the intersection of this Set and Set `a`.")(setHr(TTHr), setHr(TTHr), setHr(TTHr))
   registerMethod("difference", (x: Set[Any], a: Set[Any]) => x &~ a, "Returns the elements of this Set that are not in Set `a`.")(setHr(TTHr), setHr(TTHr), setHr(TTHr))
-  registerMethod("issubset", (x: Set[Any], a: Set[Any]) => x.subsetOf(a), "Returns true if this Set is a subset of Set `a`.")(setHr(TTHr), setHr(TTHr), boolHr)
+  registerMethod("isSubset", (x: Set[Any], a: Set[Any]) => x.subsetOf(a), "Returns true if this Set is a subset of Set `a`.")(setHr(TTHr), setHr(TTHr), boolHr)
```
Comment (cseed) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: Put this in a separate PR.

Comment (jigold) [reply src/main/scala/is/hail/expr/FunctionRegistry.scala]: See #2250

File: src/main/scala/is/hail/expr/Parser.scala (hunk 7 above)
Comment (jigold) [reply src/main/scala/is/hail/expr/Parser.scala]: See #2251

File: python/hail/expr/dataset.py (hunk 6 above)
Comment (jigold) [reply python/hail/expr/dataset.py]: I copied this from the old `dataset.py` file. @tpoterba did you add this line?

File: python/hail/expr/__init__.py (hunk 3 above)
Comment (jigold) [reply python/hail/expr/__init__.py]: I was thinking it would be this:
```
from hail2 import *
import hail2.expr.functions as hf

hc = HailContext()
hf.chisq(a, b, c, d)
```

My concern with automatically importing the functions is they are more likely to conflict with other packages.  I think in this case it is better to be explicit.

File: python/hail/expr/dataset.py (hunk 6 above)
Comment (tpoterba) [reply python/hail/expr/dataset.py]: Yeah, I added it. Specifically, in #1297. 

I did some archaeology -- I added it because I used warnings to warn about deprecation, but then killed the warnings without deleting the management line at the top. This can safely be removed.

**Comment by jigold** [comment]
@cseed I don't know why this PR isn't showing up in your stack on the PR scorecard. Can you take a second look? I know there's a lot to do still with regards to documentation and porting over methods, but I'd like to get what I have into master and progress from there.

**Comment by jigold** [comment]
Never mind. Guess there's a delay on the PR scorecard.

**Comment by cseed** [comment]
Can you fix the build?  I'd like to look over the docs before I approve.

**Comment by jigold** [comment]
Done.

**Comment by jigold** [comment]
@cseed I can't find the comment you are referring to. What do I need to modify?

**Comment by cseed** [comment]
Looks my comment got lost!  Sorry.

I said, I'd prefer we didn't copy the HailContext whole hog, but just write a simple wrapper that calls from hail2.HailContext to hail.HailContext, so something like:

```
class HailContext:
  def __init__(args...):
    self.hc1 = hail.HailContext(args...)

  def import_bgen(args...):
    return self.hc1.import_bgen(args...).to_hail2()
```

etc.  I don't think you even need docs unless there is something specifically different between the two.  That way, we won't need to maintain two versions for things like doc changes and we won't get confused about which one is the "real" HailContext.  Then, when we're ready to switch over, we can pull the docs across and throw away the stub.
